nwplus/livesite
# User status and new applicant creation + routing overhaul

### User status
This PR adds a user status field to the user object provided by auth. You can access the user status like this
```javascript
user.status // this is the users status
```
The list of available status' are in the Constants.js file under applicantStatus.
### New app creation
When a new user is detected (E.G. no application found for that user) the portal will automatically create a new document with a few fields already set.:
1. The first field set is the email, firstName and lastName fields in BasicInfo
2. The second field set is the _id field which matches the document id as well as the users id.
3. a timestamp will be set for when the document was created

### New redirections
- After logging in, you will be redirected to the correct place. This means `/application` for most statuses, but if we detect that the application status is set to `attending` then you will be redirected to `/judging`.
- Navigating to `/login` while already logged in will now redirect you to the correct place

### Auth loading / routing overhaul
- There was an issue where our authentication would load slightly after our routes did, causing the application to kind of jump around as it receives new information. To fix this, I have added a loading screen for all routes that require authentication and they won't be loaded until auth is finished loading
 **note: I plan to use the same blocking mechanism to also load whether the application is open or not in a later PR**
- Routing has been overhauled because I don't want to block loading of the portal on routes that don't require it. 
   - All Auth routes have been moved to a route at the end of the routing table
   - All Auth routes have been wrapped inside the AuthProvider component
   - Unfortunate it means the 404 component also had to be moved into this route, so it will require a quick load before displaying. (*any suggestions welcome on how to avoid this*)

@panjenny0 You can just stick your loading component inside `auth.js` where the `<>loading..</>` part is

### Testing
- Run this locally (auth doesn't work on previews)
- navigate to either the login page or any of the protected routes (/judging, /application, etc)
- Login, you should be redirected to the `/application` page
- login to firebase and make sure that a new document has been created for you under `/Hackathons/LHD2021/Applicants/{uuid}`
   - Also make sure that your firstName, lastName, email, and _id are set correctly.

Let me know if it works for you and if you have any issues with the routing overhaul.
